### ISSUE TITLE
docker uid/gid mapping when userns is configured in docker daemon.

### ISSUE THREAD

**Issue by coldbloodx** [moby/moby#48130]
### Description

Hey docker/moby developpers,

We've hit some trouble on mounting user storage after we enabled docker userns mode.

see below cases:
env:
```
[leo@lanode01 ~]$ docker --version
Docker version 26.1.3, build b72abbb
[leo@lanode01 ~]$ docker info  |grep 'Security Options' -A 5
 Security Options:
  seccomp
   Profile: builtin
  userns
  cgroupns
 Kernel Version: 5.14.0-427.18.1.el9_4.x86_64

[leo@lanode01 ~]$ id 
uid=1000(leo) gid=1000(leo) groups=1000(leo),991(docker)
```

test  case 1:
```
[leo@lanode01 ~]$ docker run -v /home/leo:/home/leo  -it $ubi  bash
[root@d4816a343dcb /]# ls -ld /home/leo
drwx------ 11 nobody nobody 4096 Jul  4 06:43 /home/leo
[root@d4816a343dcb /]# cd /home/leo/
bash: cd: /home/leo/: Permission denied
```

test case 2:
```
[leo@lanode01 ~]$ docker run -u `id -u leo`:`id -g leo` -v /home/leo:/home/leo  -it $ubi  bash
bash-5.1$ id
uid=1000 gid=1000 groups=1000
bash-5.1$ ls -l /home/leo/
ls: cannot open directory '/home/leo/': Permission denied

```

on RHEL/Centos based distro default user home directory permission: **700**
on debian/Ubuntu based distro default user home directory permisson: **750**

As we can see, if user enables docker user ns for security, user cannot access his/her home directory except change his/her home dir permission to let other users have **read** permission.
**I DO NOT think this make any sense :).**

Several weeks ago, I tried podman, and userns=keep-id caught my attention, with userns=keep-id, it sovles above user home directory problem.
see below cases:
```
[leo@lanode01 ~]$ podman run --userns=keep-id -v /home/leo:/home/leo -it $ubi  bash
[leo@3a7054c32bef /]$ ls -ld /home/leo
drwx------ 11 leo leo 4096 Jul  4 07:16 /home/leo
[leo@3a7054c32bef /]$ id
uid=1000(leo) gid=1000(leo) groups=1000(leo)
[leo@3a7054c32bef /]$ cd /home/leo/; ls
go
```

now my question/request here is:
**Is it possible for the docker team to port this feature to docker , or develop a similar, or more convenient one?**

**Comment by coldbloodx** [comment]
any comments on this one?

**Comment by Clockwork-Muse** [comment]
Duplicate of https://github.com/moby/moby/issues/42134?

Among other issues.

**Comment by cpuguy83** [comment]
This is a known problem that is not easily fixed and is currently working "as expected" even though we know the current behavior is not ideal.

See https://github.com/moby/moby/issues/2259 for more history here.
The TLDR of ^ is the UID/GID on your files are the same regardless of userns mappings.
The uesrns mappings do not change the actual file permissions and up until very recently it required a userspace filesystem (such as bindfs) to do the mapping.
On newer kernels it is possible to remap the uid/gids on files there are some issues here:

1. file system support, other than overlayfs, didn't hit until kernel 6.8 (afaict), which makes this a no-go for bind mounts except on *very* recent kernels.
2. mapping can only happen once in a namespace (??? memory is a little fuzzy here)
3. Requires changing a whole new kernel mount API

Basically, the pieces are coming into place but it is still going to take more time.

I'm going to close this because it is effectively a dup of #2259

**Comment by polarathene** [comment]
> 1. file system support, other than overlayfs, didn't hit until kernel 6.8 (afaict), which makes this a no-go for bind mounts except on _very_ recent kernels.

Wouldn't ID mapped mounts would be appropriate for this functionality? That's been a thing since kernel 5.12 (April 2021)

Your team seems to be aware of this though so I must be misunderstanding 😅

---

## References

- [Kernel Docs](https://docs.kernel.org/filesystems/idmappings.html#idmappings-on-idmapped-mounts).
- [Support merged into containerd in Sep 2023](https://github.com/containerd/containerd/pull/5890), intended for [part of the upcoming containerd 2.0 release](https://github.com/containerd/containerd/issues/9310#issuecomment-1786033104).
- Also seems to be [supported in podman](https://github.com/containerd/containerd/pull/5890#issuecomment-1270804528).

---

https://man.archlinux.org/man/mount_setattr.2.en#NOTES

> ![image](https://github.com/user-attachments/assets/7b096eb8-48b3-43a9-97a8-22f434599eb7)

> ![image](https://github.com/user-attachments/assets/64e76570-9113-442f-8c5d-02788cdf6010)

---

You use idmapped mounts with upper/lower layers of OverlayFS like described here: https://github.com/containers/podman/issues/10374#issuecomment-1559630334

There seems to be plenty of filesystem support as shown above, so I'm not sure what you're referring to about kernel 6.8 🤔

I've been assuming that sometime after Docker adopts containerd 2.0 (_once it's released_), that it'd make it easier to add support for volumes to use ID mapped mounts?

**Comment by cpuguy83** [comment]
@polarathene I must have misread the doc (or interpreted a search result, which seemed to return 6.8, incorrectly). Indeed I had to look it up because last time I reviewed there was _extremely_ limited support.

**Comment by coldbloodx** [comment]
@cpuguy83 
If @polarathene is correct and kernel 5.12 starts supporting mapping bind mounts, 
docker team will add this feature soon, right?

**Comment by polarathene** [comment]
> docker team will add this feature soon, right?

Need to wait for containerd 2.0 release AFAIK, and then a release of Docker that brings that in to use the related API.

I haven't investigated it further, so I'm not sure what the actual scope of work looks like. There may still be other blockers / caveats, but the reference links I've provided above should at least help when someone has time to investigate further 👍

**Comment by coldbloodx** [comment]
@polarathene 
thank u for the answer :)
